variable/test/variable_special_values_test.cpp

```diff
@@ -0,0 +1,211 @@
+#include <gtest/gtest.h>
+#include <vector>
+
+#include "fix_typed_test_suite_warnings.h"
+#include "test_macros.h"
+
+#include "scipp/core/element/special_values.h"
+#include "scipp/variable/arithmetic.h"
+#include "scipp/variable/special_values.h"
+#include "scipp/variable/variable.h"
+
+using namespace scipp;
+using namespace scipp::core;
+using namespace scipp::variable;
+
+template <typename T> constexpr auto values_for_special_value_tests() {
+  return std::vector{static_cast<T>(0.0),
+                     static_cast<T>(-1.23),
+                     static_cast<T>(3e4),
+                     std::numeric_limits<T>::quiet_NaN(),
+                     std::numeric_limits<T>::signaling_NaN(),
+                     std::numeric_limits<T>::infinity(),
+                     -std::numeric_limits<T>::infinity()};
+}
+
+template <typename T>
+class VariableSpecialValueTest : public ::testing::Test {};
+using FloatTypes = ::testing::Types<double, float>;
+TYPED_TEST_SUITE(VariableSpecialValueTest, FloatTypes);
+
+TYPED_TEST(VariableSpecialValueTest, isnan) {
+  for (TypeParam x : values_for_special_value_tests<TypeParam>()) {
+    for (auto u : {units::dimensionless, units::m}) {
+      const auto res = isnan(x * u);
+      EXPECT_EQ(res.template value<bool>(), element::isnan(x));
+      EXPECT_EQ(res.unit(), units::dimensionless);
+    }
+  }
```

Comment:
The modern way (you will still find some "legacy" tests providing bad examples) of writing tests for variables is to use `==` on variables, rather than their contents:

```cpp
EXPECT_EQ(isnan(x * units::m), element::isnan(x) * units::one);
```

Also note that re-testing for different units and types is not required for "trivial" cases where we simply call `transform`, since (a) transform is tested, and (b) the kernel is tested. This is to reduce the amount of test code and maintenance burden, but of course we need to find the balance in every case (and there *are* more complex operations such as `sum` that are not a simple call to `transform`).